pkg/gadgets/internal/tchandler/tc.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !withoutebpf
+
+package tchandler
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/cilium/ebpf"
+	"github.com/florianl/go-tc"
+	tccore "github.com/florianl/go-tc/core"
+	"golang.org/x/sys/unix"
+)
+
+func ptr[T any](v T) *T {
+	return &v
+}
+
+type AttachmentSense int
+
+const (
+	SenseUnspec AttachmentSense = iota
+	SenseIngress
+	SenseEgress
+)
+
+// createClsActQdisc creates a clsact qdisc on the given interface. If the qdisc already exists it
+// returns nil.
+func createClsActQdisc(tcnl *tc.Tc, iface *net.Interface) (*tc.Object, error) {
+	// Install Qdisc on interface
+	qdisc := &tc.Object{
+		Msg: tc.Msg{
+			Family:  unix.AF_UNSPEC,
+			Ifindex: uint32(iface.Index),
+			Handle:  tccore.BuildHandle(tc.HandleRoot, 0x0000),
+			Parent:  tc.HandleIngress,
+		},
+		Attribute: tc.Attribute{
+			Kind: "clsact",
+		},
+	}
+
+	if err := tcnl.Qdisc().Add(qdisc); err != nil && !errors.Is(err, unix.EEXIST) {
+		return nil, fmt.Errorf("adding clsact qdisc to %s: %w", iface.Name, err)
+	}
+
+	return qdisc, nil
+}
+
+// addTCFilter adds a filter to the given interface. It returns the filter object or an error. In
+// order to allow multiple filters on the same interface, it tries to add the filter with different
+// handles until it succeeds or it has tried 128 times.
+func addTCFilter(tcnl *tc.Tc, prog *ebpf.Program, iface *net.Interface, typ AttachmentSense) (*tc.Object, error) {
+	info, _ := prog.Info()
+	var parent uint32
+
+	switch typ {
+	case SenseIngress:
+		parent = tccore.BuildHandle(tc.HandleRoot, tc.HandleMinIngress)
+	case SenseEgress:
+		parent = tccore.BuildHandle(tc.HandleRoot, tc.HandleMinEgress)
+	default:
+		return nil, fmt.Errorf("invalid filter type")
+	}
+
+	for handle := uint32(0x1); handle < 128; handle++ {
```

Comment:
I chose 128 because:
- It's big enough to support a big quantity of programs attached (128).
- It's small enough not to loose a lot of time trying different handles in case they're all already taken.

Reply:
I don't think it's something we want to expose to users. My idea is to experiment with this value and then increase it (or use a different logic, like using different priorities with the same handle, use a random generated handle, etc.) later on if it doesn't work